fix(tm2): deduct gas fee after runTx

onlyhyde opened the pull request:
<!-- please provide a detailed description of the changes made in this pull request. -->
WIP
addresses #1070 #1067 
<details><summary>Contributors' checklist...</summary>

- [ ] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ ] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
</details>

notJoon reviewed (approved):
In the meeting, I received confirmation that the concerning issues cannot occur. LGTM

remove: `review/triage-pending` flag

onlyhyde commented:
Update as new issues are discovered.

### Issue:
After processing a trasaction, if there are not enough assets for the gas fee, it is treated as a failure, but the status change is applied.

### Cause:
After executing the runMsgs function in the runTx function, `msCache.MultiWrite()` updates the state if there is no failure in the result.
The gas fee collection is processed afterward, returning a failure, but the state remains updated.

### Fix:
- Change to collect the gas fee before updating the status.
- If an error occurs while collecting the gas fee, return an error.
- If you're processing a genesis block, add a condition to not collect the gas fee.

Gno2D2 commented:
I'm a bot that assists the Gno Core team in maintaining this repository. My role is to ensure that contributors understand and follow our guidelines, helping to streamline the development process.

The following requirements must be fulfilled before a pull request can be merged.
Some requirement checks are automated and can be verified by the CI, while others need manual verification by a staff member.

These requirements are defined in this [configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

## Automated Checks

 🔴 Maintainers must be able to edit this pull request
 🔴 The pull request head branch must be up-to-date with its base


## Manual Checks

- [ ] The pull request description provides enough details


<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>The pull request head branch must be up-to-date with its base</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 Head branch (gnoswap-labs:fix/deduct-gasfee) is up to date with base (master): behind by 27 / ahead by 8

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>The pull request description provides enough details</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 Not (🔴 Pull request author is a member of the team: core-contributors)

```
### Can be checked by
 - team core-contributors

</blockquote></details>

</blockquote></details>

</blockquote></details>

piux2 commented:
@onlyhyde Thank you so much for your contribution!

The dynamic gas price implementation is available at:
https://github.com/gnolang/gno/pull/2838

It addresses the following issues:

https://github.com/gnolang/gno/issues/1070
https://github.com/gnolang/gno/issues/1067
Please let me know if you think differently.

As mentioned in this comment:
https://github.com/gnolang/gno/pull/3209#issuecomment-2505196735
could you open a new issue if the problem still exists? Additionally, could you share the test cases and the results?

piux2 commented:
@onlyhyde  

We deduct the GasFee upfront in the AnteHandler, not after runTx

One purpose of charging the gas fee is to prevent abuse, as there is a cost associated with executing a transaction even if it fails. If we were to charge the fee only after the transaction has already been executed, it would defeat this purpose.

Kouteki commented:
@piux2 do we treat current state as a feature, and not a bug? And close this PR?

piux2 commented:
> @piux2 do we treat current state as a feature, and not a bug? And close this PR?
I don't think this is a bug, and we could close it. However, let's wait for @onlyhyde's feedback before proceeding.

Additionally, I'd like to know from @onlyhyde whether below is an issue with the master branch or with this PR.
https://github.com/gnolang/gno/pull/3209#issuecomment-2505196735

onlyhyde commented:
@piux2 

Hi, sorry for the late response.
The #3209 comment has already been fixed, the first implementation was collecting gas after runTx, then I found the case in #3209, made the fix, and changed it to collect gas fee before updating the status, not after runTx. I added the #3209 comment to share the details, as I've committed the fix in the current PR, and I've updated it further.
As you said, it has already been modified to a different PR, so the current PR can be closed.
